### ISSUE TITLE
undefined method `group_id' due to failure to create VPC security group

### ISSUE THREAD

**Issue by msaffitz** [infochimps-labs/ironfan#238]
From #232 

@temujin9 said:

> 498f0c9 should fix the Elastic IP in VPC bug. The other (undefined method `group_id' ) 
> seems to be an issue with earlier failure to actually create the VPC security group in 
> question, and will require separate fix; please file a separate issue.

**Comment by jcf** [comment]
I think I'm experiencing this same error, which is preventing me from provisioning a fresh cluster.

```
WARNING: undefined method `group_id' for nil:NilClass
ERROR: undefined method `group_id' for nil:NilClass (NoMethodError)
…
WARNING: Error launching #<NoMethodError: undefined method `group_id' for nil:NilClass>; skipping after-launch tasks.
```
- Full backtrace https://gist.github.com/4666349
- Ironfan v4.7.6
- Chef v10.18.2

**Comment by temujin9** [comment]
Hey guys. I've bumped this in importance, but we don't actively use VPC at the moment, so it may be a bit before we can manage an internal fix. I'll gladly test and accept any pull requests, if you have the urge to try.

**Comment by jcf** [comment]
I'm not using VPC so my issue may be unrelated.

**Comment by temujin9** [comment]
Yeah, I think it's something separate. Could you put this in a new issue, and include the whole context of the launch run and the cluster file you're using?

**Comment by mrflip** [comment]
Choose one:

a) the security group isn't made / discovered, and therefore this is
correctly crashing (though perhaps it would be nice if it said why). In
this case, we should add a better error message but not allow the machine
to launch
b) As long as ironfan prominently warns, I'd rather launch my machine and
tidy security groups. Fixing the crash to a warning is useful

If b), I pushed a workaround in a side branch for you to try out.

flip

**Comment by temujin9** [comment]
I dislike (b), especially since you still can't (AFAIK) change security groups in non-VPC instances.

**Comment by jcf** [comment]
I think the fault might be my own.

I've started with the example web cluster and reached the following configuration by commenting out sections that I think are unnecessary for my current purposes.

https://gist.github.com/4668103
- Does Ironfan create security groups that don't exist?
- Does Ironfan modify a security group if the ports listed change?

@mrflip thank you for pushing the patch. If I understand the code correctly it warns that a security group can not be found to avoid the exception caused by calling `group_id` on a non-existent security group.

**Comment by temujin9** [comment]
Ironfan should create nonexistent groups, and add to port openings (but not remove old ones), anytime it syncs: any `knife cluster sync`, `knife cluster launch`, or `knife cluster start` command, and probably others as well.

**Comment by jcf** [comment]
Hmm. That's perplexing then.

Is there a simple way to test my knife credentials to make sure it's not an authorisation issue?

---

When I run `knife cluster show spoonfeed` I get:

```
Inventorying servers in spoonfeed cluster, all facets, all servers
  spoonfeed:        Loading chef
  spoonfeed:        Loading ec2
  spoonfeed:        Reconciling DSL and provider information
  +---------------------+-------+-------------+-----------+------------+-----+
  | Name                | Chef? | State       | Flavor    | AZ         | Env |
  +---------------------+-------+-------------+-----------+------------+-----+
  | spoonfeed-webnode-0 | yes   | not running | m1.medium | eu-west-1a | dev |
  +---------------------+-------+-------------+-----------+------------+-----+
```

**Comment by temujin9** [comment]
First thing to check is whether the launch run tried to make the groups in question. Run your launch, but include `-V`. (Due to a questionable design decision on bundler's part, you probably have to do `bundle exec "knife cluster launch -V"` to get the verbosity flag through to the right place). That should tell you what it's doing without too much noise. If not, use `-VV`; that'll be hideously noisy, so capturing it to a file might be good.

If those don't immediately highlight the problem, gist the results and open another issue.

**Comment by jcf** [comment]
Ahem. This is a little embarrassing. If I `bundle exec knife cluster …` everything works as expected.

Thank you very much for your help.

**Comment by temujin9** [comment]
Okay, that tells me that the bundler stuff needs to be made more transparent. I've seen a way to accomplish that in a Rakefile somewhere; opening an issue to fix this. Thanks for running into that wall enough to make it visible . . .

**Comment by temujin9** [comment]
https://github.com/infochimps-labs/ironfan/issues/246 is the issue, if you want to keep an eye on progress.

**Comment by jcf** [comment]
Thanks Nathaniel. :thumbsup:

**Comment by temujin9** [comment]
Okay, unfortunately making the bundle inclusion run inside of the knife plugin introduces more problems than it solves. I've opened #253 instead; closing this, as the proximal issue has been addressed.

**Comment by garo** [comment]
I'm just starting with Ironfan and I'm trying to provision a test server into VPC and I ran into the same problem:

<pre>
WARNING: Error running #<Ironfan::Broker::Computer(server=#<Ironfan::Dsl::Server(name="0", run_list_items=c{ role[systemwide], role[ssh], role[set_hostname], log_integration::logrotate, role[volumes], role[package_set], role[minidash], role[org_base], role[org_users], role[org_final], role[tuning], role[sandbox-cluster], role[sandbox-simple-facet] }, clouds=c{ ec2 }, volumes=c{  }, environment=:development, cluster_role=#<Ironfan::Dsl::Role>, facet_role=#<Ironfan::Dsl::Role>, cluster_name="sandbox", facet_name="simple")>, resources=c{ client, node, role__sandbox-cluster, role__sandbox-simple-facet, keypair, security_group__vpc-69dfb507:default }, drives=c{ root }, providers=c{ chef, iaas })>:
WARNING: undefined method `group_id' for nil:NilClass
ERROR: undefined method `group_id' for nil:NilClass (NoMethodError)
/opt/chef/embedded/lib/ruby/gems/1.9.1/gems/ironfan-4.12.3/lib/ironfan/provider/ec2/machine.rb:260:in `block in launch_description'
/opt/chef/embedded/lib/ruby/gems/1.9.1/gems/ironfan-4.12.3/lib/ironfan/provider/ec2/machine.rb:259:in `map'
/opt/chef/embedded/lib/ruby/gems/1.9.1/gems/ironfan-4.12.3/lib/ironfan/provider/ec2/machine.rb:259:in `launch_description'
/opt/chef/embedded/lib/ruby/gems/1.9.1/gems/ironfan-4.12.3/lib/ironfan/provider/ec2/machine.rb:167:in `create!'
/opt/chef/embedded/lib/ruby/gems/1.9.1/gems/ironfan-4.12.3/lib/ironfan/broker/computer.rb:77:in `launch'
/opt/chef/embedded/lib/ruby/gems/1.9.1/gems/ironfan-4.12.3/lib/ironfan/broker/computer.rb:316:in `block in launch'
/opt/chef/embedded/lib/ruby/gems/1.9.1/gems/ironfan-4.12.3/lib/ironfan.rb:42:in `block (3 levels) in parallel'
/opt/chef/embedded/lib/ruby/gems/1.9.1/gems/ironfan-4.12.3/lib/ironfan.rb:155:in `safely'
/opt/chef/embedded/lib/ruby/gems/1.9.1/gems/ironfan-4.12.3/lib/ironfan.rb:41:in `block (2 levels) in parallel'
ERROR: /opt/chef/embedded/lib/ruby/gems/1.9.1/gems/ironfan-4.12.3/lib/ironfan/provider/ec2/machine.rb:260:in `block in launch_description'
/opt/chef/embedded/lib/ruby/gems/1.9.1/gems/ironfan-4.12.3/lib/ironfan/provider/ec2/machine.rb:259:in `map'
/opt/chef/embedded/lib/ruby/gems/1.9.1/gems/ironfan-4.12.3/lib/ironfan/provider/ec2/machine.rb:259:in `launch_description'
/opt/chef/embedded/lib/ruby/gems/1.9.1/gems/ironfan-4.12.3/lib/ironfan/provider/ec2/machine.rb:167:in `create!'
/opt/chef/embedded/lib/ruby/gems/1.9.1/gems/ironfan-4.12.3/lib/ironfan/broker/computer.rb:77:in `launch'
/opt/chef/embedded/lib/ruby/gems/1.9.1/gems/ironfan-4.12.3/lib/ironfan/broker/computer.rb:316:in `block in launch'
/opt/chef/embedded/lib/ruby/gems/1.9.1/gems/ironfan-4.12.3/lib/ironfan.rb:42:in `block (3 levels) in parallel'
/opt/chef/embedded/lib/ruby/gems/1.9.1/gems/ironfan-4.12.3/lib/ironfan.rb:155:in `safely'
/opt/chef/embedded/lib/ruby/gems/1.9.1/gems/ironfan-4.12.3/lib/ironfan.rb:41:in `block (2 levels) in parallel'
WARNING: Error launching #<NoMethodError: undefined method `group_id' for nil:NilClass>; skipping after-launch tasks.
</pre>

I'm running ironfan 4.12.3 as you can see. Here's my slightly modified sandbox.rb:

<pre>
Ironfan.cluster 'sandbox' do
  cloud(:ec2) do
    permanent           false
    availability_zones ['us-east-1c']
    flavor              't1.micro'
    backing             'ebs'
    image_name          'ironfan-precise'
    bootstrap_distro    'ubuntu12.04-ironfan'
    chef_client_script  'client.rb'
    mount_ephemerals
    vpc                 'vpc-69dfb507'
    subnet              'subnet-ea9bd584'
  end
  environment           :development
  role                  :systemwide,    :first
  role                  :ssh
  cloud(:ec2).security_group(:default).authorize_port_range 22..22
  role                  :set_hostname
  recipe                'log_integration::logrotate'
  role                  :volumes
  role                  :package_set,   :last
  role                  :minidash,      :last
  role                  :org_base
  role                  :org_users
  role                  :org_final,     :last
  role                  :tuning,        :last
  facet :simple do
    instances           1
  end
  cluster_role.override_attributes({
    })
end
</pre>

**Comment by garo** [comment]
I did further analysis on the code and I found the root problem. The file ironfan/provider/ec2/security_group.rb contains function def self.ensure_groups(computer) which seems to ensure that there's security group with the name cluster (sandbox) and for the each facet (sandbox-simple), but it doesn't call authorized_by_group() for the facet security group.

This results that def self.prepare!(computers) does not add the facet security group to the groups_to_create list, thus the sandbox-facet security group is missing and the later code will fail.
